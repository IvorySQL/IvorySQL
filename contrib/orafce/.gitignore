*.o
*.a
*.so
*.so.*
*.sdf
*.opensdf
*.suo
*.*.user
/.deps/
/orafce.sql
/orafce.sql.in
/results
[Oo]bj/
[Bb]in/