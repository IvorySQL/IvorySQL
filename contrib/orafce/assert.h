#ifndef __ASSERT__
#define __ASSERT__

#define ERRCODE_ORA_PACKAGES_INVALID_SCHEMA_NAME	MAKE_SQLSTATE('4','4','0','0','1')
#define ERRCODE_ORA_PACKAGES_INVALID_OBJECT_NAME	MAKE_SQLSTATE('4','4','0','0','2')
#define ERRCODE_ORA_PACKAGES_ISNOT_SIMPLE_SQL_NAME	MAKE_SQLSTATE('4','4','0','0','3')
#define ERRCODE_ORA_PACKAGES_ISNOT_QUALIFIED_SQL_NAME	MAKE_SQLSTATE('4','4','0','0','4')

#endif
