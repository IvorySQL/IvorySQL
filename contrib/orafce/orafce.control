# orafce extension
comment = 'Functions and operators that emulate a subset of functions and packages from the Oracle RDBMS'
default_version = '3.17'
module_pathname = '$libdir/orafce'
relocatable = false
