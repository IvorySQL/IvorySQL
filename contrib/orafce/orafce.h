#ifndef __ORAFCE__
#define __ORAFCE__

#include "postgres.h"
#include "catalog/catversion.h"
#include "nodes/pg_list.h"
#include <sys/time.h>
#include "utils/datetime.h"
#include "utils/datum.h"

#define TextPCopy(t) \
	DatumGetTextP(datumCopy(PointerGetDatum(t), false, -1))

#define PG_GETARG_IF_EXISTS(n, type, defval) \
	((PG_NARGS() > (n) && !PG_ARGISNULL(n)) ? PG_GETARG_##type(n) : (defval))

/* alignment of this struct must fit for all types */
typedef union vardata
{
	char	c;
	short	s;
	int		i;
	long	l;
	float	f;
	double	d;
	void   *p;
} vardata;

extern int ora_instr(text *txt, text *pattern, int start, int nth);
extern int ora_mb_strlen(text *str, char **sizes, int **positions);
extern int ora_mb_strlen1(text *str);

extern char *nls_date_format;
extern char *orafce_timezone;

extern bool orafce_varchar2_null_safe_concat;

/*
 * Version compatibility
 */

extern Oid	equality_oper_funcid(Oid argtype);

/*
 * Date utils
 */
#define STRING_PTR_FIELD_TYPE const char *const

extern STRING_PTR_FIELD_TYPE ora_days[];

extern int ora_seq_search(const char *name, STRING_PTR_FIELD_TYPE array[], size_t max);

#ifdef _MSC_VER

#define int2size(v)			(size_t)(v)
#define size2int(v)			(int)(v)

#else

#define int2size(v)			v
#define size2int(v)			v

#endif

#endif
