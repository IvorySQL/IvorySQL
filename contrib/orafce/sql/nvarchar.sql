/* type nvarchar test */

set compatible_mode to oracle;

create table nvarchar_t1 (id int, name nvarchar(32));
insert into nvarchar_t1 values (1,'ab'),(2,'as');
create table nvarchar_t2 (id int, name nchar(32));
insert into nvarchar_t2 values (1,'ab'),(2,'as');
select * from nvarchar_t1 where name = 'ab';
select * from nvarchar_t1 where name = 'as';
select * from nvarchar_t1 where name = 'ab     ';
select * from nvarchar_t1 where name = 'ab'or name = 'as';
select * from nvarchar_t1 where name = 'ab'or name = 'as';
select nvarchar_t1.name || nvarchar_t2.name from nvarchar_t1,nvarchar_t2;
select * from nvarchar_t1 where name <> 'ab';
select * from nvarchar_t1 where name <> 'cc';

select * from nvarchar_t1 where name > 'ac';
select * from nvarchar_t1 where name >= 'ac';
select * from nvarchar_t1 where name < 'ac';
select * from nvarchar_t1 where name <= 'ac';

select * from nvarchar_t1 where name ~ 'ac';
select * from nvarchar_t1 where name ~ 'ab';
select * from nvarchar_t1 where name ~ 'a';
select * from nvarchar_t1 where name !~ 'ac';
select * from nvarchar_t1 where name !~ 'ab';
select * from nvarchar_t1 where name !~ 'a';

select * from nvarchar_t1 where name ~~ 'ab';
select * from nvarchar_t1 where name !~~ 'ab';
select * from nvarchar_t1 where name ~* 'ab';
select * from nvarchar_t1 where name !~* 'ab';
select * from nvarchar_t1 where name ~~* 'ab';
select * from nvarchar_t1 where name !~~* 'ab';
select * from nvarchar_t1 where name ~<~ 'ab';
select * from nvarchar_t1 where name ~<=~ 'ab';
select * from nvarchar_t1 where name ~>~ 'ab';
select * from nvarchar_t1 where name ~>=~ 'ab';

create table varchar_t11 (id int, name varchar(32));
insert into varchar_t11 values (1,'ab'),(2,'as');
create table varchar_t22 (id int, name char(32));
insert into varchar_t22 values (1,'ab'),(2,'as');
select * from varchar_t11 where name = 'ab';
select * from varchar_t11 where name = 'as';
select * from varchar_t11 where name = 'ab     ';
select * from varchar_t11 where name = 'ab'or name = 'as';
select * from varchar_t11 where name = 'ab'or name = 'as';
select varchar_t11.name || varchar_t22.name from varchar_t11,varchar_t22;
select * from varchar_t11 where name <> 'ab';
select * from varchar_t11 where name <> 'cc';

select * from varchar_t11 where name > 'ac';
select * from varchar_t11 where name >= 'ac';
select * from varchar_t11 where name < 'ac';
select * from varchar_t11 where name <= 'ac';

select * from varchar_t11 where name ~ 'ac';
select * from varchar_t11 where name ~ 'ab';
select * from varchar_t11 where name ~ 'a';
select * from varchar_t11 where name !~ 'ac';
select * from varchar_t11 where name !~ 'ab';
select * from varchar_t11 where name !~ 'a';

select * from varchar_t11 where name ~~ 'ab';
select * from varchar_t11 where name !~~ 'ab';
select * from varchar_t11 where name ~* 'ab';
select * from varchar_t11 where name !~* 'ab';
select * from varchar_t11 where name ~~* 'ab';
select * from varchar_t11 where name !~~* 'ab';
select * from varchar_t11 where name ~<~ 'ab';
select * from varchar_t11 where name ~<=~ 'ab';
select * from varchar_t11 where name ~>~ 'ab';
select * from varchar_t11 where name ~>=~ 'ab';

drop table nvarchar_t1, varchar_t11;

create or replace function nvarchartestfunc1(str nvarchar)
returns nvarchar as $$
begin
	return 'this is a nvarchar type, value: ' || str;
end;
$$ language plpgsql;

create or replace function nvarchartestfunc11(str varchar)
returns varchar as $$
begin
	return 'this is a varchar type, value: ' || str;
end;
$$ language plpgsql;

create table tb_text (name text);
insert into tb_text values ('haha');
select nvarchartestfunc1(name) from tb_text;
drop table tb_text;

create table tb_bpchar (name char(32));
insert into tb_bpchar values ('haha');
select nvarchartestfunc1(name) from tb_bpchar;
drop table tb_bpchar;

create table tb_char (name "char");
insert into tb_char values ('A');
select nvarchartestfunc1(name) from tb_char;
select nvarchartestfunc11(name) from tb_char;
drop table tb_char;

create table tb_varchar (name varchar(32));
insert into tb_varchar values ('haha');
select nvarchartestfunc1(name) from tb_varchar;
drop table tb_varchar;

create table tb_name (name name);
insert into tb_name values ('haha');
select nvarchartestfunc1(name) from tb_name;
select nvarchartestfunc11(name) from tb_name;
drop table tb_name;

create table tb_bool (name bool);
insert into tb_bool values ('true');
select nvarchartestfunc1(name) from tb_bool;
select nvarchartestfunc11(name) from tb_bool;
drop table tb_bool;

create table tb_varchar2 (name varchar2);
insert into tb_varchar2 values ('haha');
select nvarchartestfunc1(name) from tb_varchar2;
drop table tb_varchar2;

create table tb_nvarchar2 (name nvarchar2);
insert into tb_nvarchar2 values ('haha');
select nvarchartestfunc1(name) from tb_nvarchar2;
drop table tb_nvarchar2;

drop FUNCTION nvarchartestfunc1;
drop FUNCTION nvarchartestfunc11;
