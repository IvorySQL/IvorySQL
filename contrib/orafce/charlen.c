/*
 * charlen.c
 *
 * provides a modified version of bpcharlen() that does not
 * ignore trailing spaces of CHAR arguments to provide an
 * Oracle compatible length() function
 */

#include "postgres.h"

#include "utils/builtins.h"
#include "access/hash.h"
#include "libpq/pqformat.h"
#include "nodes/nodeFuncs.h"
#include "utils/array.h"
#include "utils/formatting.h"
#include "mb/pg_wchar.h"
#include "fmgr.h"

#include "orafce.h"
#include "builtins.h"

PG_FUNCTION_INFO_V1(orafce_bpcharlen);

Datum
orafce_bpcharlen(PG_FUNCTION_ARGS)
{
	BpChar     *arg = PG_GETARG_BPCHAR_PP(0);
	int         len;

	/* byte-length of the argument (trailing spaces not ignored) */
	len = VARSIZE_ANY_EXHDR(arg);

	/* in multibyte encoding, convert to number of characters */
	if (pg_database_encoding_max_length() != 1)
		len = pg_mbstrlen_with_len(VARDATA_ANY(arg), len);

	PG_RETURN_INT32(len);
}
