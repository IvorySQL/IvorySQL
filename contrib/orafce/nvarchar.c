
/*----------------------------------------------------------------------------
 *
 *     nvarchar.c
 *     nvarchar type for PostgreSQL.
 *
 *----------------------------------------------------------------------------
 */

#include "postgres.h"
#include "access/hash.h"
#include "libpq/pqformat.h"
#include "nodes/nodeFuncs.h"
#include "utils/array.h"
#include "utils/builtins.h"
#include "utils/guc.h"
#include "mb/pg_wchar.h"
#include "fmgr.h"

#include "orafce.h"
#include "builtins.h"


PG_FUNCTION_INFO_V1(nvarcharin);
PG_FUNCTION_INFO_V1(nvarchar);

/*
 * varchar_input -- common guts of varcharin and varcharrecv
 *
 * s is the input text of length len (may not be null-terminated)
 * atttypmod is the typmod value to apply
 *
 * Note that atttypmod is measured in characters, which
 * is not necessarily the same as the number of bytes.
 *
 * If the input string is too long, raise an error, unless the extra
 * characters are spaces, in which case they're truncated.  (per SQL)
 *
 * Uses the C string to text conversion function, which is only appropriate
 * if VarChar and text are equivalent types.
 */
static VarChar *
nvarchar_input(const char *s, size_t len, int32 atttypmod)
{
	VarChar    *result;
	size_t		maxlen;

	maxlen = atttypmod - VARHDRSZ;

	if (atttypmod >= (int32) VARHDRSZ && len > maxlen)
	{
		/* Verify that extra characters are spaces, and clip them off */
		size_t		mbmaxlen = pg_mbcharcliplen(s, len, maxlen);
		size_t		j;

		for (j = mbmaxlen; j < len; j++)
		{
			if (s[j] != ' ')
				ereport(ERROR,
						(errcode(ERRCODE_STRING_DATA_RIGHT_TRUNCATION),
						 errmsg("value too long for type character varying(%d)",
								(int) maxlen)));
		}

		len = mbmaxlen;
	}

	result = (VarChar *) cstring_to_text_with_len(s, len);
	return result;
}

/*
 * Convert a C string to VARCHAR internal representation.  atttypmod
 * is the declared length of the type plus VARHDRSZ.
 */
Datum
nvarcharin(PG_FUNCTION_ARGS)
{
	char	   *s = PG_GETARG_CSTRING(0);

#ifdef NOT_USED
	Oid			typelem = PG_GETARG_OID(1);
#endif
	int32		atttypmod = PG_GETARG_INT32(2);
	VarChar    *result;

	result = nvarchar_input(s, strlen(s), atttypmod);
	PG_RETURN_VARCHAR_P(result);
}


/*
 * nvarcharout()
 * converts a NVARCHAR value to a C string.
 *
 * Uses the text to C string conversion function, which is only appropriate
 * if VarChar and text are equivalent types.
 *
 * just use varcharout()
 */


/*
 * nvarcharrecv()
 * converts external binary format to nvarchar
 *
 * just use varcharrecv()
 */


/*
 * nvarcharsend -- convert nvarchar to binary value
 *
 * just use varcharsend()
 */


/*
 * Converts a VARCHAR type to the specified size.
 *
 * maxlen is the typmod, ie, declared length plus VARHDRSZ bytes.
 * isExplicit is true if this is for an explicit cast to varchar(N).
 *
 * Truncation rules: for an explicit cast, silently truncate to the given
 * length; for an implicit cast, raise error unless extra characters are
 * all spaces.  (This is sort-of per SQL: the spec would actually have us
 * raise a "completion condition" for the explicit cast case, but Postgres
 * hasn't got such a concept.)
 */
Datum
nvarchar(PG_FUNCTION_ARGS)
{
	VarChar    *source = PG_GETARG_VARCHAR_PP(0);
	int32		typmod = PG_GETARG_INT32(1);
	bool		isExplicit = PG_GETARG_BOOL(2);
	int32		len,
				maxlen;
	size_t		maxmblen;
	int			i;
	int			charlen;
	char	   *s_data;
	int			perstrlen = 0;

	len = VARSIZE_ANY_EXHDR(source);
	s_data = VARDATA_ANY(source);
	maxlen = typmod - VARHDRSZ;

	/* No work if typmod is invalid or supplied data fits it already */
	if (maxlen < 0 || len <= maxlen)
		PG_RETURN_VARCHAR_P(source);

	/* only reach here if string is too long... */

	/* truncate multibyte string preserving multibyte boundary */
	maxmblen = pg_mbcharcliplen(s_data, len, maxlen);

	charlen = pg_mbstrlen_with_len(s_data, len);

	if (!isExplicit)
	{
		for (i = maxmblen; i < len; i++)
			if (s_data[i] != ' ')
				ereport(ERROR,
						(errcode(ERRCODE_STRING_DATA_RIGHT_TRUNCATION),
						 errmsg("value too long for type character varying(%d)",
								maxlen)));
	}

	/* Added these because 'Chinese character'::varchar(5) to 'Chinese character'::varchar(3),
	 * display first Chinese character.
	 */
	if (!isExplicit)
	{
		perstrlen = pg_mblen(s_data);
		if (perstrlen > 1)
		{
			int iscomplate = 0;
			bool iscomplate1 = false;

			if (charlen * perstrlen == len)
				iscomplate1 = true;

			iscomplate = maxlen % perstrlen;

			if (iscomplate1 == true)
				maxmblen = maxlen - iscomplate;
			else if (maxlen < maxmblen)
				maxmblen = maxlen;
		}
	}

	PG_RETURN_VARCHAR_P((VarChar *) cstring_to_text_with_len(s_data,
															 maxmblen));
}


/*
 * nvarchartypmodin -- type modifier input function
 *
 * just use varchartypmodin()
 */

/*
 * nvarchartypmodout -- type modifier output function
 *
 * just use varchartypmodout()
 */
