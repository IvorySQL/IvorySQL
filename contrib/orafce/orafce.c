#include "postgres.h"
#include "storage/lwlock.h"
#include "storage/shmem.h"
#include "utils/guc.h"
#include "commands/variable.h"

#include "orafce.h"
#include "builtins.h"
#include "pipe.h"

/*  default value */
char  *nls_date_format = NULL;
char  *orafce_timezone = NULL;

void
_PG_init(void)
{

#if PG_VERSION_NUM < 90600

	RequestAddinLWLocks(1);

#endif

	RequestAddinShmemSpace(SHMEMMSGSZ);

	/* Define custom GUC variables. */
	DefineCustomStringVariable("orafce.nls_date_format",
									"Emulate oracle's date output behaviour.",
									NULL,
									&nls_date_format,
									NULL,
									PGC_USERSET,
									0,
									NULL,
									NULL, NULL);

	DefineCustomStringVariable("orafce.timezone",
									"Specify timezone used for sysdate function.",
									NULL,
									&orafce_timezone,
									"GMT",
									PGC_USERSET,
									0,
									check_timezone, NULL, NULL);

	DefineCustomBoolVariable("orafce.varchar2_null_safe_concat",
									"Specify timezone used for sysdate function.",
									NULL,
									&orafce_varchar2_null_safe_concat,
									false,
									PGC_USERSET,
									0,
									NULL, NULL, NULL);

	EmitWarningsOnPlaceholders("orafce");
}
