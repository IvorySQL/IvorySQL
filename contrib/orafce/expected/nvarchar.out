/* type nvarchar test */
set compatible_mode to oracle;
create table nvarchar_t1 (id int, name nvarchar(32));
insert into nvarchar_t1 values (1,'ab'),(2,'as');
create table nvarchar_t2 (id int, name nchar(32));
insert into nvarchar_t2 values (1,'ab'),(2,'as');
select * from nvarchar_t1 where name = 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name = 'as';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name = 'ab     ';
 id | name 
----+------
(0 rows)

select * from nvarchar_t1 where name = 'ab'or name = 'as';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from nvarchar_t1 where name = 'ab'or name = 'as';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select nvarchar_t1.name || nvarchar_t2.name from nvarchar_t1,nvarchar_t2;
 ?column? 
----------
 abab
 abas
 asab
 asas
(4 rows)

select * from nvarchar_t1 where name <> 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name <> 'cc';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from nvarchar_t1 where name > 'ac';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name >= 'ac';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name < 'ac';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name <= 'ac';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name ~ 'ac';
 id | name 
----+------
(0 rows)

select * from nvarchar_t1 where name ~ 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name ~ 'a';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from nvarchar_t1 where name !~ 'ac';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from nvarchar_t1 where name !~ 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name !~ 'a';
 id | name 
----+------
(0 rows)

select * from nvarchar_t1 where name ~~ 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name !~~ 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name ~* 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name !~* 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name ~~* 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name !~~* 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name ~<~ 'ab';
 id | name 
----+------
(0 rows)

select * from nvarchar_t1 where name ~<=~ 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from nvarchar_t1 where name ~>~ 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from nvarchar_t1 where name ~>=~ 'ab';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

create table varchar_t11 (id int, name varchar(32));
insert into varchar_t11 values (1,'ab'),(2,'as');
create table varchar_t22 (id int, name char(32));
insert into varchar_t22 values (1,'ab'),(2,'as');
select * from varchar_t11 where name = 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name = 'as';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name = 'ab     ';
 id | name 
----+------
(0 rows)

select * from varchar_t11 where name = 'ab'or name = 'as';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from varchar_t11 where name = 'ab'or name = 'as';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select varchar_t11.name || varchar_t22.name from varchar_t11,varchar_t22;
 ?column? 
----------
 abab
 abas
 asab
 asas
(4 rows)

select * from varchar_t11 where name <> 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name <> 'cc';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from varchar_t11 where name > 'ac';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name >= 'ac';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name < 'ac';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name <= 'ac';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name ~ 'ac';
 id | name 
----+------
(0 rows)

select * from varchar_t11 where name ~ 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name ~ 'a';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from varchar_t11 where name !~ 'ac';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

select * from varchar_t11 where name !~ 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name !~ 'a';
 id | name 
----+------
(0 rows)

select * from varchar_t11 where name ~~ 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name !~~ 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name ~* 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name !~* 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name ~~* 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name !~~* 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name ~<~ 'ab';
 id | name 
----+------
(0 rows)

select * from varchar_t11 where name ~<=~ 'ab';
 id | name 
----+------
  1 | ab
(1 row)

select * from varchar_t11 where name ~>~ 'ab';
 id | name 
----+------
  2 | as
(1 row)

select * from varchar_t11 where name ~>=~ 'ab';
 id | name 
----+------
  1 | ab
  2 | as
(2 rows)

drop table nvarchar_t1, varchar_t11;
create or replace function nvarchartestfunc1(str nvarchar)
returns nvarchar as $$
begin
	return 'this is a nvarchar type, value: ' || str;
end;
$$ language plpgsql;
create or replace function nvarchartestfunc11(str varchar)
returns varchar as $$
begin
	return 'this is a varchar type, value: ' || str;
end;
$$ language plpgsql;
create table tb_text (name text);
insert into tb_text values ('haha');
select nvarchartestfunc1(name) from tb_text;
          nvarchartestfunc1           
--------------------------------------
 this is a nvarchar type, value: haha
(1 row)

drop table tb_text;
create table tb_bpchar (name char(32));
insert into tb_bpchar values ('haha');
select nvarchartestfunc1(name) from tb_bpchar;
          nvarchartestfunc1           
--------------------------------------
 this is a nvarchar type, value: haha
(1 row)

drop table tb_bpchar;
create table tb_char (name "char");
insert into tb_char values ('A');
select nvarchartestfunc1(name) from tb_char;
ERROR:  function nvarchartestfunc1("char") does not exist
LINE 1: select nvarchartestfunc1(name) from tb_char;
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
select nvarchartestfunc11(name) from tb_char;
ERROR:  function nvarchartestfunc11("char") does not exist
LINE 1: select nvarchartestfunc11(name) from tb_char;
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
drop table tb_char;
create table tb_varchar (name varchar(32));
insert into tb_varchar values ('haha');
select nvarchartestfunc1(name) from tb_varchar;
          nvarchartestfunc1           
--------------------------------------
 this is a nvarchar type, value: haha
(1 row)

drop table tb_varchar;
create table tb_name (name name);
insert into tb_name values ('haha');
select nvarchartestfunc1(name) from tb_name;
ERROR:  function nvarchartestfunc1(name) does not exist
LINE 1: select nvarchartestfunc1(name) from tb_name;
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
select nvarchartestfunc11(name) from tb_name;
ERROR:  function nvarchartestfunc11(name) does not exist
LINE 1: select nvarchartestfunc11(name) from tb_name;
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
drop table tb_name;
create table tb_bool (name bool);
insert into tb_bool values ('true');
select nvarchartestfunc1(name) from tb_bool;
ERROR:  function nvarchartestfunc1(boolean) does not exist
LINE 1: select nvarchartestfunc1(name) from tb_bool;
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
select nvarchartestfunc11(name) from tb_bool;
ERROR:  function nvarchartestfunc11(boolean) does not exist
LINE 1: select nvarchartestfunc11(name) from tb_bool;
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
drop table tb_bool;
create table tb_varchar2 (name varchar2);
insert into tb_varchar2 values ('haha');
select nvarchartestfunc1(name) from tb_varchar2;
          nvarchartestfunc1           
--------------------------------------
 this is a nvarchar type, value: haha
(1 row)

drop table tb_varchar2;
create table tb_nvarchar2 (name nvarchar2);
insert into tb_nvarchar2 values ('haha');
select nvarchartestfunc1(name) from tb_nvarchar2;
          nvarchartestfunc1           
--------------------------------------
 this is a nvarchar type, value: haha
(1 row)

drop table tb_nvarchar2;
drop FUNCTION nvarchartestfunc1;
drop FUNCTION nvarchartestfunc11;
